### ISSUE TITLE
[Discussion] zod2 - default() condition

### ISSUE THREAD

**Issue by tmtron** [vriad/zod#162]
The current implementation replaces all falsy values to the default.   
I was surprised by this. I expected that only `undefined` is checked - but maye I'm just biased because I am used to Joi: [Joi.default()](https://joi.dev/api/?v=17.2.1#anydefaultvalue)

There are also cases where we don't want this.  
* a number may have a default of `5`, but `0` could also be a valid value
* and also `null` can sometimes be a valid value (different from the default)

The options that we have:  
1. `default()` could have a parameter to decide which to use
1. we could have 3 different functions: `defaultIfFalsy()`, `defaultIfUndefined()`, `defaultIfUndefinedOrNull()`
1. or just document in detail how `default()` function works and force the users of the lib to create custom transformers for the other cases

**Comment by marlonbernardes** [comment]
Good point - I think `.default()` should behave similarly to how the null coalescing operator does: if a value is either null or undefined it should use the default value. If it's `0`, an empty string, an empty array or `false` a default value shouldn't be used.

Regardless of how it's implemented I think it would be nice to add this to the documentation. I'm happy to contribute with both the code and documentation changes.

**Comment by marlonbernardes** [comment]
One more example on why using the default when falsy might not be a good idea: booleans cannot be set to false if you use `.default()`:

```ts
const schema = z.object({
  sandbox: z.boolean().default(true)
});

// Good - returns { sandbox: true } - which makes sense as it's not supplied
schema.parse({}); 

// Bad! - returns { sandbox: true } even though I've provided a  value for it.
schema.parse({ sandbox: false }); 
```

**Comment by vriad** [comment]
Fixed in zod@2.0.0-beta.9

Good catch, this was a dumb mistake on my part. The default value should only be applied if the input is `undefined`.
